### ISSUE TITLE
The validation action loss does not go down  

### ISSUE THREAD

**Issue by hk-zh** [lukashermann/hulc#5]
First of all, thank you for your excellent work. Your provided policy has a very high success rate. 
When training with your hulc module, I found that the training loss went down smoothly. However, the validation loss does not go down since the first epoch. The following figure shows my action_loss_pp on wandb. 
<img width="341" alt="Screenshot 2022-11-05 at 22 20 42" src="https://user-images.githubusercontent.com/57254021/200141765-8fac6cd0-393e-4654-9253-b1bc9ab9c769.png">
I trained with dataset D's training set and validated with D's validation set. 
I checked the validation step and I think it goes well. I don't know if its the problem of D's validation set

**Comment by lukashermann** [comment]
We experienced the same and it is not informative of the performance of a trained policy. If you want an accurate metric of how well your policy performs, you would have to run the evaluation rollouts (or rather a subset of them) during training, like every 5 epochs. 
As an example, I can show you this old run where the success rate of `chain_5_tasks` goes up nicely while the validation loss goes up

![W B Chart 11_7_2022, 9_56_37 AM](https://user-images.githubusercontent.com/23167230/200269160-a4206809-8793-437f-846e-c05eb9c0f2b4.png)

![W B Chart 11_7_2022, 9_56_18 AM](https://user-images.githubusercontent.com/23167230/200269127-d2713a8e-ccd3-4e9e-8e3c-2dab232e1b0f.png)

**Comment by hk-zh** [comment]
Ok, I got it. Do you have some clues what is the reason for this happening? Could I also ask how many epochs you trained your hulc model to get over 80% success rate?

**Comment by lukashermann** [comment]
With 80% success rate you mean the success rate for completing the first subtask? We trained for around 30 epochs with 8 GPUs and a batch size of 32.

**Comment by hk-zh** [comment]
Thanks for the information. It really helps!

**Comment by lukashermann** [comment]
For clarification, with batch size 32 I mean the batch size per GPU.

**Comment by hk-zh** [comment]
> For clarification, with batch size 32 I mean the batch size per GPU.

Yap, got it.

**Comment by Cherryjingyao** [comment]
Besides  the  success rate  ,can I just make the training loss as clue of the mode? when trained at 30 epochs the loss go down at what value?
I ask the question  because i did not the the evaluate rollout which bring the crash as I mentioned in ALSA lib issue.

**Comment by lukashermann** [comment]
I don't completely understand your question. There are multiple things that are informative of the performance apart from the evaluation rollouts. You can visualize the latent distribution of prior and posterior as a TSNE plot and see if the tasks are clustered nicely. In the validation step, comparing lang_act_loss_pr and lang_act_loss_pp gives you an idea of how well the prior can replace the posterior at test time wrt the action generation. But ultimately the evaluation rollouts are much more informative.

**Comment by Cherryjingyao** [comment]
Because I can not run the eval rollout during the training , I'm curious about how to decided how well the policy perform without eval rollout
you mentioned lang_act_loss_pr and lang_act_loss_pp ,can val_kl_loss also be informative?
but I found the val_kl_loss did not go down during training.
<img width="1071" alt="image" src="https://github.com/lukashermann/hulc/assets/30522880/76e1f2b5-5a87-4208-ad4b-8cb1ab7b9df2">

**Comment by lukashermann** [comment]
I'm afraid I don't have a good answer for you. It doesn't necessarily need to be a problem if the loss goes up since you're training the plan recognition network (posterior) also with the action loss and the KL loss serves as both a regularizer for the plan recognition, as well as it trains the prior to approximate the posterior. I recommend you to spend some time getting the eval rollouts to work during training.
